shotover-proxy/src/transforms/cassandra/peers_rewrite.rs

```diff
@@ -24,50 +27,76 @@ impl CassandraPeersRewriteConfig {
 #[derive(Clone)]
 pub struct CassandraPeersRewrite {
     port: u32,
+    peer_table: FQName,
 }
 
 impl CassandraPeersRewrite {
     pub fn new(port: u32) -> Self {
-        CassandraPeersRewrite { port }
+        CassandraPeersRewrite {
+            port,
+            peer_table: FQName::new("system", "peers_v2"),
+        }
     }
 }
 
 #[async_trait]
 impl Transform for CassandraPeersRewrite {
     async fn transform<'a>(&'a mut self, mut message_wrapper: Wrapper<'a>) -> ChainResponse {
         // Find the indices of queries to system.peers & system.peers_v2
-        let system_peers: Vec<usize> = message_wrapper
+        // we need to know which columns in which CQL queries in which messages have system peers
+        let column_names: Vec<(usize, Vec<Identifier>)> = message_wrapper
             .messages
             .iter_mut()
             .enumerate()
-            .filter_map(|(i, m)| if is_system_peers(m) { Some(i) } else { None })
+            .filter_map(|(i, m)| {
+                let sys_peers = extract_native_port_column(&self.peer_table, m);
+                if sys_peers.is_empty() {
+                    None
+                } else {
+                    Some((i, sys_peers))
+                }
+            })
             .collect();
 
         let mut response = message_wrapper.call_next_transform().await?;
 
-        for i in system_peers {
-            rewrite_port(&mut response[i], self.port);
+        for (i, name_list) in column_names {
+            rewrite_port(&mut response[i], &name_list, self.port);
         }
 
         Ok(response)
     }
 }
 
-fn is_system_peers(message: &mut Message) -> bool {
-    if let Some(Frame::Cassandra(_)) = message.frame() {
-        if let Some(namespace) = message.namespace() {
-            if namespace.len() > 1 {
-                return namespace[0] == "system" && namespace[1] == "peers_v2";
+/// determine if the message contains a SELECT from `system.peers_v2` that includes the `native_port` column
+/// return a list of column names (or their alias) for each `native_port`.
+fn extract_native_port_column(peer_table: &FQName, message: &mut Message) -> Vec<Identifier> {
```

Comment:
This will need some perf revalidation, @conorbros had to spend some time getting the fast path (e.g. exit early if it's not a query we are interested in) right previously

Reply:
I dont think this transform ever had a fast path.
A proper fast path would need to skip query parsing and nothing of that sort was ever implemented.
Maybe you are thinking of one of the many other cassandra optimizations conor did e.g. https://github.com/shotover/shotover-proxy/pull/637

A fast path would be useful but we should investigate as a follow up to this PR.

Reply:
+1 re follow up PR

Reply:
cool I'm happy with that